File: 2502. Design Memory Allocator/Solution.py
from collections import defaultdict


class Allocator:

    def __init__(self, n: int):
        self.mem_table = {0: n}
        self.id2blocks = defaultdict(list)

    def allocate(self, size: int, mID: int) -> int:
        beg_addr = -1
        addrs = sorted(self.mem_table.keys())
        for addr in addrs:
            if self.mem_table[addr] >= size:
                beg_addr = addr
                break
        if beg_addr == -1:
            return -1
        self.id2blocks[mID].append((beg_addr, size))
        if self.mem_table[beg_addr] > size:
            self.mem_table[beg_addr + size] = self.mem_table[beg_addr] - size
        self.mem_table.pop(beg_addr)
        return beg_addr

    def freeMemory(self, mID: int) -> int:
        if mID not in self.id2blocks:
            return 0
        blocks = self.id2blocks.pop(mID)
        ans = 0
        for addr, blk_size in blocks:
            ans += blk_size
            self.mem_table[addr] = blk_size

        del_addrs = set()
        addrs = sorted(self.mem_table.keys())
        for addr in addrs:
            while addr not in del_addrs and addr + self.mem_table[addr] in self.mem_table:
                del_addrs.add(addr + self.mem_table[addr])
                self.mem_table[addr] += self.mem_table[addr + self.mem_table[addr]]
        for addr in del_addrs:
            self.mem_table.pop(addr)
        return ans


# Your Allocator object will be instantiated and called as such:
# obj = Allocator(n)
# param_1 = obj.allocate(size,mID)
# param_2 = obj.freeMemory(mID)

File: 2502. Design Memory Allocator/Solution.cpp
#include <map>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

class Allocator {
public:
    Allocator(int n)
    {
        mem_table[0] = n;
    }

    int allocate(int size, int mID)
    {
        int beg_addr = -1;
        for (auto [k, v] : mem_table) {
            if (v >= size) {
                beg_addr = k;
                break;
            }
        }
        if (beg_addr == -1)
            return -1;
        id2blocks[mID].emplace_back(beg_addr, size);
        if (mem_table[beg_addr] > size)
            mem_table[beg_addr + size] = mem_table[beg_addr] - size;
        mem_table.erase(beg_addr);
        return beg_addr;
    }

    int freeMemory(int mID)
    {
        vector<pair<int, int>>& blocks = id2blocks[mID];
        int ans = 0;
        for (const pair<int, int>& block : blocks) {
            ans += block.second;
            auto it = mem_table.emplace(block.first, block.second).first;

            if (it != mem_table.begin()) {
                auto prev_it = prev(it);
                if (prev_it->first + prev_it->second == block.first) {
                    prev_it->second += block.second;
                    mem_table.erase(it);
                    it = prev_it;
                }
            }

            auto next_it = next(it);
            if (next_it != mem_table.end() && it->first + it->second == next_it->first) {
                it->second += next_it->second;
                mem_table.erase(next_it);
            }
        }
        id2blocks.erase(mID);
        return ans;
    }

private:
    map<int, int> mem_table;
    unordered_map<int, vector<pair<int, int>>> id2blocks;
};

/**
 * Your Allocator object will be instantiated and called as such:
 * Allocator* obj = new Allocator(n);
 * int param_1 = obj->allocate(size,mID);
 * int param_2 = obj->freeMemory(mID);
 */
